File: src/TaskOTime/TaskOTime.ViewModel/ViewModels/TaskManagementViewModel.vb
Imports System
Imports System.ComponentModel
Imports System.Collections.ObjectModel
Imports System.Linq
Imports System.Windows.Input
Imports TaskOTime.ViewModel.Base

Namespace ViewModels
    Public Class TaskManagementViewModel
        Inherits ViewModelBase

        Private _selectedTaskList As TaskListViewModel
        Private _selectedTaskItem As TaskItemViewModel
        Private _currentRecordingTask As TaskItemViewModel
        Private _completeRecordingOnNextTimeEntry As Boolean
        Private ReadOnly _newTaskCommand As DelegateCommand
        Private ReadOnly _editTaskCommand As DelegateCommand
        Private ReadOnly _newTaskListCommand As DelegateCommand
        Private ReadOnly _editTaskListCommand As DelegateCommand
        Private ReadOnly _startTaskCommand As DelegateCommand
        Private ReadOnly _completeTaskCommand As DelegateCommand
        Private ReadOnly _moreToDoCommand As DelegateCommand

        Public Sub New()
            TaskLists = CreateSampleTaskLists()

            _newTaskCommand = New DelegateCommand(Sub(parameter) AddTask())
            _editTaskCommand = New DelegateCommand(Sub(parameter) EditSelectedTask(), Function(parameter) SelectedTaskItem IsNot Nothing)
            _newTaskListCommand = New DelegateCommand(Sub(parameter) RequestNewTaskList())
            _editTaskListCommand = New DelegateCommand(Sub(parameter) RequestEditTaskList(), Function(parameter) SelectedTaskList IsNot Nothing)
            _startTaskCommand = New DelegateCommand(Sub(parameter) StartSelectedTask(), Function(parameter) CanStartSelectedTask())
            _completeTaskCommand = New DelegateCommand(Sub(parameter) CompleteSelectedTask(), Function(parameter) CanCompleteSelectedTask())
            _moreToDoCommand = New DelegateCommand(Sub(parameter) MarkSelectedTaskNeedsMore(), Function(parameter) CanMarkSelectedTaskNeedsMore())

            SelectedTaskList = TaskLists.FirstOrDefault()
        End Sub

        Public Event TaskListEditRequested As EventHandler(Of TaskListEditRequestEventArgs)

        Public ReadOnly Property TaskLists As ObservableCollection(Of TaskListViewModel)

        Public Property SelectedTaskList As TaskListViewModel
            Get
                Return _selectedTaskList
            End Get
            Set(value As TaskListViewModel)
                If SetProperty(_selectedTaskList, value, NameOf(SelectedTaskList)) Then
                    If value Is Nothing Then
                        SelectedTaskItem = Nothing
                    Else
                        SelectedTaskItem = value.Tasks.FirstOrDefault(Function(taskItem) taskItem.IsOpen)
                        If SelectedTaskItem Is Nothing Then
                            SelectedTaskItem = value.Tasks.FirstOrDefault()
                        End If
                    End If

                    OnPropertyChanged(NameOf(SelectedListSummary))
                    RaiseCommandStatesChanged()
                End If
            End Set
        End Property

        Public Property SelectedTaskItem As TaskItemViewModel
            Get
                Return _selectedTaskItem
            End Get
            Set(value As TaskItemViewModel)
                If SetProperty(_selectedTaskItem, value, NameOf(SelectedTaskItem)) Then
                    OnPropertyChanged(NameOf(SelectedTaskSummary))
                    RaiseCommandStatesChanged()
                End If
            End Set
        End Property

        Public Property CompleteRecordingOnNextTimeEntry As Boolean
            Get
                Return _completeRecordingOnNextTimeEntry
            End Get
            Set(value As Boolean)
                SetProperty(_completeRecordingOnNextTimeEntry, value, NameOf(CompleteRecordingOnNextTimeEntry))
            End Set
        End Property

        Public ReadOnly Property CurrentRecordingTask As TaskItemViewModel
            Get
                Return _currentRecordingTask
            End Get
        End Property

        Public ReadOnly Property IsTaskRecording As Boolean
            Get
                Return CurrentRecordingTask IsNot Nothing AndAlso CurrentRecordingTask.IsStarted
            End Get
        End Property

        Public ReadOnly Property RecordingStatusText As String
            Get
                If Not IsTaskRecording Then
                    Return String.Empty
                End If

                Return CurrentRecordingTask.RecordingStatusText
            End Get
        End Property

        Public ReadOnly Property NewTaskCommand As ICommand
            Get
                Return _newTaskCommand
            End Get
        End Property

        Public ReadOnly Property EditTaskCommand As ICommand
            Get
                Return _editTaskCommand
            End Get
        End Property

        Public ReadOnly Property NewTaskListCommand As ICommand
            Get
                Return _newTaskListCommand
            End Get
        End Property

        Public ReadOnly Property EditTaskListCommand As ICommand
            Get
                Return _editTaskListCommand
            End Get
        End Property

        Public ReadOnly Property StartTaskCommand As ICommand
            Get
                Return _startTaskCommand
            End Get
        End Property

        Public ReadOnly Property CompleteTaskCommand As ICommand
            Get
                Return _completeTaskCommand
            End Get
        End Property

        Public ReadOnly Property MoreToDoCommand As ICommand
            Get
                Return _moreToDoCommand
            End Get
        End Property

        Public ReadOnly Property TaskPanelTitle As String
            Get
                Return "Aufgaben"
            End Get
        End Property

        Public ReadOnly Property TaskPanelSummary As String
            Get
                Return "Listen oben, Aufgaben unten – vorbereitet für spätere AppServer-Daten."
            End Get
        End Property

        Public ReadOnly Property SelectedListSummary As String
            Get
                If SelectedTaskList Is Nothing Then
                    Return "Keine Aufgabenliste ausgewählt."
                End If

                Return SelectedTaskList.ProgressSummary
            End Get
        End Property

        Public ReadOnly Property SelectedTaskSummary As String
            Get
                If SelectedTaskItem Is Nothing Then
                    Return "Keine Aufgabe ausgewählt."
                End If

                Return SelectedTaskItem.ActionHint
            End Get
        End Property

        Private Shared Function CreateSampleTaskLists() As ObservableCollection(Of TaskListViewModel)
            Return New ObservableCollection(Of TaskListViewModel) From {
                New TaskListViewModel(
                    "Mein Tag",
                    "Fokus für heute",
                    {
                        New TaskItemViewModel("Zeitbuchungen prüfen", "Tageszeiten kurz mit dem Kalender abgleichen.", "Heute"),
                        New TaskItemViewModel("Rückfragen klären", "Offene Punkte aus der Morgenrunde nachfassen.", "Heute"),
                        New TaskItemViewModel("Tagesabschluss vorbereiten", "Notizen und offene Blöcke für den Feierabend sammeln.", "17:00")
                    }),
                New TaskListViewModel(
                    "Projektarbeit",
                    "Aktive Liefergegenstände",
                    {
                        New TaskItemViewModel("UI-Schnitt abstimmen", "Aufgabenpanel mit Zeiterfassung verzahnen.", "Diese Woche"),
                        New TaskItemViewModel("Service-Vertrag skizzieren", "Felder für spätere AppServer-Tasks festhalten.", "Später")
                    }),
                New TaskListViewModel(
                    "Nachverfolgung",
                    "Wartet auf Antwort",
                    {
                        New TaskItemViewModel("Freigabe einholen", "Status beim Fachbereich nachfragen.", "Morgen"),
                        New TaskItemViewModel("Dokumentation ergänzen", "Kurze Bediennotiz für Aufgabenstatus vormerken.", "Diese Woche")
                    })
            }
        End Function

        Private Sub AddTask()
            If SelectedTaskList Is Nothing Then
                Return
            End If

            Dim taskItem = New TaskItemViewModel("Neue Aufgabe", "Beschreibung ergänzen.", "Heute")
            AddHandler taskItem.PropertyChanged, AddressOf OnRecordingTaskPropertyChanged
            SelectedTaskList.Tasks.Add(taskItem)
            SelectedTaskItem = taskItem
            RefreshSelectedTaskState()
        End Sub

        Private Sub EditSelectedTask()
            If SelectedTaskItem Is Nothing Then
                Return
            End If

            SelectedTaskItem.Title = "Bearbeitet: " & SelectedTaskItem.Title
            SelectedTaskItem.Description = "Aufgabe wurde über die vorbereitete Bearbeiten-Aktion geändert."
            RefreshSelectedTaskState()
        End Sub

        Private Sub RequestNewTaskList()
            RaiseEvent TaskListEditRequested(
                Me,
                New TaskListEditRequestEventArgs(
                    "Neue Aufgabenliste",
                    "Beschreibung ergänzen",
                    Sub(title, subtitle)
                        Dim taskList = New TaskListViewModel(title, subtitle, Enumerable.Empty(Of TaskItemViewModel)())
                        TaskLists.Add(taskList)
                        SelectedTaskList = taskList
                    End Sub))
        End Sub

        Private Sub RequestEditTaskList()
            If SelectedTaskList Is Nothing Then
                Return
            End If

            RaiseEvent TaskListEditRequested(
                Me,
                New TaskListEditRequestEventArgs(
                    SelectedTaskList.Title,
                    SelectedTaskList.Subtitle,
                    Sub(title, subtitle)
                        SelectedTaskList.Title = title
                        SelectedTaskList.Subtitle = subtitle
                    End Sub))
        End Sub

        Private Function CanStartSelectedTask() As Boolean
            Return SelectedTaskItem IsNot Nothing AndAlso
                   Not SelectedTaskItem.IsDone AndAlso
                   Not SelectedTaskItem.IsStarted
        End Function

        Private Sub StartSelectedTask()
            If Not CanStartSelectedTask() Then
                Return
            End If

            If CurrentRecordingTask IsNot Nothing AndAlso CurrentRecordingTask IsNot SelectedTaskItem Then
                CurrentRecordingTask.StopRecordingWithoutFinishing()
                ClearCurrentRecording()
            End If

            SelectedTaskItem.MarkStarted()
            _currentRecordingTask = SelectedTaskItem
            AddHandler _currentRecordingTask.PropertyChanged, AddressOf OnRecordingTaskPropertyChanged
            OnPropertyChanged(NameOf(CurrentRecordingTask))
            OnPropertyChanged(NameOf(IsTaskRecording))
            OnPropertyChanged(NameOf(RecordingStatusText))
            RefreshSelectedTaskState()
        End Sub

        Private Function CanCompleteSelectedTask() As Boolean
            Return SelectedTaskItem IsNot Nothing AndAlso Not SelectedTaskItem.IsDone
        End Function

        Private Sub CompleteSelectedTask()
            If Not CanCompleteSelectedTask() Then
                Return
            End If

            SelectedTaskItem.MarkDone()
            ClearCurrentRecording()
            RefreshSelectedTaskState()
        End Sub

        Private Function CanMarkSelectedTaskNeedsMore() As Boolean
            Return SelectedTaskItem IsNot Nothing AndAlso Not SelectedTaskItem.IsDone
        End Function

        Private Sub MarkSelectedTaskNeedsMore()
            If Not CanMarkSelectedTaskNeedsMore() Then
                Return
            End If

            SelectedTaskItem.MarkNeedsMore()
            ClearCurrentRecording()
            RefreshSelectedTaskState()
        End Sub

        Public Sub StopRecordingAfterTimeEntry(completeTask As Boolean)
            If CurrentRecordingTask Is Nothing Then
                Return
            End If

            If completeTask Then
                CurrentRecordingTask.MarkDone()
            Else
                CurrentRecordingTask.StopRecordingWithoutFinishing()
            End If

            CompleteRecordingOnNextTimeEntry = False
            ClearCurrentRecording()
            RefreshSelectedTaskState()
        End Sub

        Public Sub UpdateRecordingClock(nowValue As DateTime)
            If CurrentRecordingTask Is Nothing Then
                Return
            End If

            CurrentRecordingTask.UpdateRecordingElapsed(nowValue)
            OnPropertyChanged(NameOf(RecordingStatusText))
        End Sub

        Private Sub ClearCurrentRecording()
            If _currentRecordingTask IsNot Nothing Then
                RemoveHandler _currentRecordingTask.PropertyChanged, AddressOf OnRecordingTaskPropertyChanged
            End If

            _currentRecordingTask = Nothing
            OnPropertyChanged(NameOf(CurrentRecordingTask))
            OnPropertyChanged(NameOf(IsTaskRecording))
            OnPropertyChanged(NameOf(RecordingStatusText))
        End Sub

        Private Sub OnRecordingTaskPropertyChanged(sender As Object, e As PropertyChangedEventArgs)
            If e.PropertyName = NameOf(TaskItemViewModel.IsStarted) OrElse
               e.PropertyName = NameOf(TaskItemViewModel.RecordingStatusText) Then
                OnPropertyChanged(NameOf(IsTaskRecording))
                OnPropertyChanged(NameOf(RecordingStatusText))
            End If
        End Sub

        Private Sub RefreshSelectedTaskState()
            OnPropertyChanged(NameOf(SelectedTaskSummary))
            OnPropertyChanged(NameOf(SelectedListSummary))
            RaiseCommandStatesChanged()
        End Sub

        Private Sub RaiseCommandStatesChanged()
            _newTaskCommand.RaiseCanExecuteChanged()
            _editTaskCommand.RaiseCanExecuteChanged()
            _newTaskListCommand.RaiseCanExecuteChanged()
            _editTaskListCommand.RaiseCanExecuteChanged()
            _startTaskCommand.RaiseCanExecuteChanged()
            _completeTaskCommand.RaiseCanExecuteChanged()
            _moreToDoCommand.RaiseCanExecuteChanged()
        End Sub
    End Class
End Namespace

File: src/TaskOTime/TaskOTime.ViewModel/ViewModels/TaskManagementViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Windows.Input;
using TaskOTime.ViewModel.Base;

namespace TaskOTime.ViewModel.ViewModels
{
    public class TaskManagementViewModel : Localization.LocalizedViewModelBase
    {

        private TaskListViewModel _selectedTaskList;
        private TaskItemViewModel _selectedTaskItem;
        private TaskItemViewModel _currentRecordingTask;
        private bool _completeRecordingOnNextTimeEntry;
        private readonly DelegateCommand _newTaskCommand;
        private readonly DelegateCommand _editTaskCommand;
        private readonly DelegateCommand _newTaskListCommand;
        private readonly DelegateCommand _editTaskListCommand;
        private readonly DelegateCommand _startTaskCommand;
        private readonly DelegateCommand _completeTaskCommand;
        private readonly DelegateCommand _moreToDoCommand;
        private readonly Func<DateTime> _clock;
        private readonly Action<TaskItemViewModel> _saveTask;

        public TaskManagementViewModel(IEnumerable<TaskListViewModel> lists = null, Func<DateTime> clock = null, Action<TaskItemViewModel> saveTask = null)
        {
            TaskLists = lists is null ? CreateSampleTaskLists() : new ObservableCollection<TaskListViewModel>(lists);
            _clock = clock ?? (() => DateTime.Now);
            _saveTask = saveTask;

            _newTaskCommand = new DelegateCommand(parameter => AddTask());
            _editTaskCommand = new DelegateCommand(parameter => EditSelectedTask(), parameter => SelectedTaskItem is not null);
            _newTaskListCommand = new DelegateCommand(parameter => RequestNewTaskList());
            _editTaskListCommand = new DelegateCommand(parameter => RequestEditTaskList(), parameter => SelectedTaskList is not null);
            _startTaskCommand = new DelegateCommand(parameter => StartSelectedTask(), parameter => CanStartSelectedTask());
            _completeTaskCommand = new DelegateCommand(parameter => CompleteSelectedTask(), parameter => CanCompleteSelectedTask());
            _moreToDoCommand = new DelegateCommand(parameter => MarkSelectedTaskNeedsMore(), parameter => CanMarkSelectedTaskNeedsMore());

            SelectedTaskList = TaskLists.FirstOrDefault();
        }

        public event EventHandler<TaskListEditRequestEventArgs> TaskListEditRequested;
        public event EventHandler<TaskCompletionRequestEventArgs> TaskCompletionRequested;

        public ObservableCollection<TaskListViewModel> TaskLists { get; private set; }

        public TaskListViewModel SelectedTaskList
        {
            get
            {
                return _selectedTaskList;
            }
            set
            {
                if (SetProperty(ref _selectedTaskList, value, nameof(SelectedTaskList)))
                {
                    if (value is null)
                    {
                        SelectedTaskItem = null;
                    }
                    else
                    {
                        SelectedTaskItem = value.Tasks.FirstOrDefault(taskItem => taskItem.IsOpen);
                        if (SelectedTaskItem is null)
                        {
                            SelectedTaskItem = value.Tasks.FirstOrDefault();
                        }
                    }

                    OnPropertyChanged(nameof(SelectedListSummary));
                    RaiseCommandStatesChanged();
                }
            }
        }

        public TaskItemViewModel SelectedTaskItem
        {
            get
            {
                return _selectedTaskItem;
            }
            set
            {
                if (SetProperty(ref _selectedTaskItem, value, nameof(SelectedTaskItem)))
                {
                    OnPropertyChanged(nameof(SelectedTaskSummary));
                    RaiseCommandStatesChanged();
                }
            }
        }

        public bool CompleteRecordingOnNextTimeEntry
        {
            get
            {
                return _completeRecordingOnNextTimeEntry;
            }
            set
            {
                SetProperty(ref _completeRecordingOnNextTimeEntry, value, nameof(CompleteRecordingOnNextTimeEntry));
            }
        }

        public TaskItemViewModel CurrentRecordingTask
        {
            get
            {
                return _currentRecordingTask;
            }
        }

        public bool IsTaskRecording
        {
            get
            {
                return CurrentRecordingTask is not null && CurrentRecordingTask.IsStarted;
            }
        }

        public string RecordingStatusText
        {
            get
            {
                if (!IsTaskRecording)
                {
                    return string.Empty;
                }

                return CurrentRecordingTask.RecordingStatusText;
            }
        }

        public ICommand NewTaskCommand
        {
            get
            {
                return _newTaskCommand;
            }
        }

        public ICommand EditTaskCommand
        {
            get
            {
                return _editTaskCommand;
            }
        }

        public ICommand NewTaskListCommand
        {
            get
            {
                return _newTaskListCommand;
            }
        }

        public ICommand EditTaskListCommand
        {
            get
            {
                return _editTaskListCommand;
            }
        }

        public ICommand StartTaskCommand
        {
            get
            {
                return _startTaskCommand;
            }
        }

        public ICommand CompleteTaskCommand
        {
            get
            {
                return _completeTaskCommand;
            }
        }

        public ICommand MoreToDoCommand
        {
            get
            {
                return _moreToDoCommand;
            }
        }

        public string TaskPanelTitle
        {
            get
            {
                return Text("Main_Tasks");
            }
        }

        public string TaskPanelSummary
        {
            get
            {
                return Text("Task_PanelSummary");
            }
        }

        public string SelectedListSummary
        {
            get
            {
                if (SelectedTaskList is null)
                {
                    return Text("Task_NoList");
                }

                return SelectedTaskList.ProgressSummary;
            }
        }

        public string SelectedTaskSummary
        {
            get
            {
                if (SelectedTaskItem is null)
                {
                    return Text("Task_NoSelection");
                }

                return SelectedTaskItem.ActionHint;
            }
        }

        private static ObservableCollection<TaskListViewModel> CreateSampleTaskLists()
        {
            return new ObservableCollection<TaskListViewModel>
            {
                new TaskListViewModel(Text("Sample_MyDay"), Text("Sample_TodayFocus"), new[]
                {
                    new TaskItemViewModel(Text("Sample_CheckBookings"), Text("Sample_CheckBookingsDescription"), Text("TodayButtonText")),
                    new TaskItemViewModel(Text("Sample_ClarifyQuestions"), Text("Sample_ClarifyQuestionsDescription"), Text("TodayButtonText")),
                    new TaskItemViewModel(Text("Sample_PrepareDayEnd"), Text("Sample_PrepareDayEndDescription"), "17:00")
                }),
                new TaskListViewModel(Text("Sample_ProjectWork"), Text("Sample_Deliverables"), new[]
                {
                    new TaskItemViewModel(Text("Sample_CoordinateUi"), Text("Sample_CoordinateUiDescription"), Text("Main_ThisWeek")),
                    new TaskItemViewModel(Text("Sample_ServiceContract"), Text("Sample_ServiceContractDescription"), Text("Sample_Later"))
                }),
                new TaskListViewModel(Text("Sample_FollowUp"), Text("Sample_Waiting"), new[]
                {
                    new TaskItemViewModel(Text("Sample_GetApproval"), Text("Sample_GetApprovalDescription"), Text("Sample_Tomorrow")),
                    new TaskItemViewModel(Text("Sample_UpdateDocumentation"), Text("Sample_UpdateDocumentationDescription"), Text("Main_ThisWeek"))
                })
            };
        }

        private void AddTask()
        {
            if (SelectedTaskList is null)
            {
                return;
            }

            var taskItem = new TaskItemViewModel(Text("NewTaskButtonText"), Text("Booking_AddDescription"), Text("TodayButtonText"));
            taskItem.PropertyChanged += OnRecordingTaskPropertyChanged;
            SelectedTaskList.Tasks.Add(taskItem);
            SelectedTaskItem = taskItem;
            RefreshSelectedTaskState();
        }

        private void EditSelectedTask()
        {
            if (SelectedTaskItem is null)
            {
                return;
            }

            SelectedTaskItem.Title = Text("Task_EditedTitle", SelectedTaskItem.Title);
            SelectedTaskItem.Description = Text("Task_EditDescription");
            RefreshSelectedTaskState();
        }

        private void RequestNewTaskList()
        {
            TaskListEditRequested?.Invoke(this, new TaskListEditRequestEventArgs(Text("NewTaskListButtonText"), Text("Booking_AddDescription"), (title, subtitle) =>
{
var taskList = new TaskListViewModel(title, subtitle, Enumerable.Empty<TaskItemViewModel>());
TaskLists.Add(taskList);
SelectedTaskList = taskList;
}));
        }

        private void RequestEditTaskList()
        {
            if (SelectedTaskList is null)
            {
                return;
            }

            TaskListEditRequested?.Invoke(this, new TaskListEditRequestEventArgs(SelectedTaskList.Title, SelectedTaskList.Subtitle, (title, subtitle) =>
{
SelectedTaskList.Title = title;
SelectedTaskList.Subtitle = subtitle;
}));
        }

        private bool CanStartSelectedTask()
        {
            return SelectedTaskItem is not null && !SelectedTaskItem.IsDone && !SelectedTaskItem.IsStarted;
        }

        private void StartSelectedTask()
        {
            if (!CanStartSelectedTask())
            {
                return;
            }

            if (CurrentRecordingTask is not null && !ReferenceEquals(CurrentRecordingTask, SelectedTaskItem))
            {
                CurrentRecordingTask.StopRecordingWithoutFinishing();
                ClearCurrentRecording();
            }

            SelectedTaskItem.MarkStarted(_clock());
            _currentRecordingTask = SelectedTaskItem;
            _currentRecordingTask.PropertyChanged += OnRecordingTaskPropertyChanged;
            OnPropertyChanged(nameof(CurrentRecordingTask));
            OnPropertyChanged(nameof(IsTaskRecording));
            OnPropertyChanged(nameof(RecordingStatusText));
            RefreshSelectedTaskState();
        }

        private bool CanCompleteSelectedTask()
        {
            return SelectedTaskItem is not null && !SelectedTaskItem.IsDone;
        }

        private void CompleteSelectedTask()
        {
            if (!CanCompleteSelectedTask())
            {
                return;
            }

            // Complete the compensating persistence request before changing the live task state.
            var task = SelectedTaskItem;
            RequestCompletion(task, _clock(), false, true);
            task.MarkDone();
            if (ReferenceEquals(CurrentRecordingTask, task))
            {
                CompleteRecordingOnNextTimeEntry = false;
                ClearCurrentRecording();
            }
            RefreshSelectedTaskState();
        }

        private bool CanMarkSelectedTaskNeedsMore()
        {
            return SelectedTaskItem is not null && !SelectedTaskItem.IsDone;
        }

        private void MarkSelectedTaskNeedsMore()
        {
            if (!CanMarkSelectedTaskNeedsMore())
            {
                return;
            }

            SelectedTaskItem.MarkNeedsMore();
            ClearCurrentRecording();
            RefreshSelectedTaskState();
        }

        public void StopRecordingAfterTimeEntry(bool completeTask, DateTime boundaryTime)
        {
            if (CurrentRecordingTask is null)
            {
                return;
            }

            var task = CurrentRecordingTask;
            RequestCompletion(task, boundaryTime, true, completeTask);
            if (completeTask)
            {
                task.MarkDone();
            }
            else
            {
                task.StopRecordingWithoutFinishing();
            }

            CompleteRecordingOnNextTimeEntry = false;
            ClearCurrentRecording();
            RefreshSelectedTaskState();
        }

        private void RequestCompletion(TaskItemViewModel task, DateTime completedAt, bool useExistingBoundary, bool completeTask)
        {
            var request = new TaskCompletionRequestEventArgs(task, completedAt, useExistingBoundary);
            try
            {
                TaskCompletionRequested?.Invoke(this, request);
                // The save callback receives the intended persisted state, but the
                // live task keeps its recording timestamp and flags until success.
                if (completeTask && _saveTask is not null)
                    _saveTask(task.CreateCompletedSnapshot());
            }
            catch (Exception failure)
            {
                try
                {
                    request.RollbackBooking?.Invoke();
                }
                catch (Exception rollbackFailure)
                {
                    throw new AggregateException(Text("Booking_CompletionRollbackFailed"), failure, rollbackFailure);
                }
                finally
                {
                    RefreshSelectedTaskState();
                }
                throw;
            }
        }

        public void UpdateRecordingClock(DateTime nowValue)
        {
            if (CurrentRecordingTask is null)
            {
                return;
            }

            CurrentRecordingTask.UpdateRecordingElapsed(nowValue);
            OnPropertyChanged(nameof(RecordingStatusText));
        }

        private void ClearCurrentRecording()
        {
            if (_currentRecordingTask is not null)
            {
                _currentRecordingTask.PropertyChanged -= OnRecordingTaskPropertyChanged;
            }

            _currentRecordingTask = null;
            OnPropertyChanged(nameof(CurrentRecordingTask));
            OnPropertyChanged(nameof(IsTaskRecording));
            OnPropertyChanged(nameof(RecordingStatusText));
        }

        private void OnRecordingTaskPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if ((e.PropertyName ?? "") == nameof(TaskItemViewModel.IsStarted) || (e.PropertyName ?? "") == nameof(TaskItemViewModel.RecordingStatusText))
            {
                OnPropertyChanged(nameof(IsTaskRecording));
                OnPropertyChanged(nameof(RecordingStatusText));
            }
        }

        private void RefreshSelectedTaskState()
        {
            OnPropertyChanged(nameof(SelectedTaskSummary));
            OnPropertyChanged(nameof(SelectedListSummary));
            RaiseCommandStatesChanged();
        }

        private void RaiseCommandStatesChanged()
        {
            _newTaskCommand.RaiseCanExecuteChanged();
            _editTaskCommand.RaiseCanExecuteChanged();
            _newTaskListCommand.RaiseCanExecuteChanged();
            _editTaskListCommand.RaiseCanExecuteChanged();
            _startTaskCommand.RaiseCanExecuteChanged();
            _completeTaskCommand.RaiseCanExecuteChanged();
            _moreToDoCommand.RaiseCanExecuteChanged();
        }
    }
}